storagenode/gracefulexit/worker.go

```diff
@@ -169,6 +122,78 @@ func (worker *Worker) Run(ctx context.Context, done func()) (err error) {
+
+	originalHash, originalOrderLimit, err := worker.getHashAndLimit(ctx, reader, addrLimit.GetLimit())
+	if err != nil {
+		worker.log.Error("failed to get piece hash and order limit.", zap.Stringer("satellite ID", worker.satelliteID), zap.Stringer("piece ID", pieceID), zap.Error(errs.Wrap(err)))
+		worker.handleFailure(ctx, pb.TransferFailed_UNKNOWN, pieceID, c.Send)
+		return err
+	}
+
+	putCtx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	// TODO what's the typical expiration setting?
+	pieceHash, peerID, err := worker.ecclient.PutPiece(putCtx, ctx, addrLimit, pk, reader, time.Now().Add(time.Second*600))
+	if err != nil {
+		if piecestore.ErrVerifyUntrusted.Has(err) {
+			worker.log.Error("failed hash verification.", zap.Stringer("satellite ID", worker.satelliteID), zap.Stringer("piece ID", pieceID), zap.Error(errs.Wrap(err)))
+			worker.handleFailure(ctx, pb.TransferFailed_HASH_VERIFICATION, pieceID, c.Send)
+		} else {
+			worker.log.Error("failed to put piece.", zap.Stringer("satellite ID", worker.satelliteID), zap.Stringer("piece ID", pieceID), zap.Error(errs.Wrap(err)))
+			// TODO look at error type to decide on the transfer error
+			worker.handleFailure(ctx, pb.TransferFailed_STORAGE_NODE_UNAVAILABLE, pieceID, c.Send)
+		}
+		return err
+	}
+
+	if !bytes.Equal(originalHash.Hash, pieceHash.Hash) {
+		worker.log.Error("piece hash from new storagenode does not match", zap.Stringer("storagenode ID", addrLimit.Limit.StorageNodeId), zap.Stringer("satellite ID", worker.satelliteID), zap.Stringer("piece ID", pieceID))
+		worker.handleFailure(ctx, pb.TransferFailed_HASH_VERIFICATION, pieceID, c.Send)
+		return Error.New("piece hash from new storagenode does not match")
+	}
+	if pieceHash.PieceId != addrLimit.Limit.PieceId {
+		worker.log.Error("piece id from new storagenode does not match order limit", zap.Stringer("storagenode ID", addrLimit.Limit.StorageNodeId), zap.Stringer("satellite ID", worker.satelliteID), zap.Stringer("piece ID", pieceID))
+		worker.handleFailure(ctx, pb.TransferFailed_HASH_VERIFICATION, pieceID, c.Send)
+		return Error.New("piece id from new storagenode does not match order limit")
+	}
+
+	signee := signing.SigneeFromPeerIdentity(peerID)
+	err = signing.VerifyPieceHashSignature(ctx, signee, pieceHash)
+	if err != nil {
+		worker.log.Error("invalid piece hash signature from new storagenode", zap.Stringer("storagenode ID", addrLimit.Limit.StorageNodeId), zap.Stringer("satellite ID", worker.satelliteID), zap.Stringer("piece ID", pieceID), zap.Error(errs.Wrap(err)))
+		worker.handleFailure(ctx, pb.TransferFailed_HASH_VERIFICATION, pieceID, c.Send)
+		return err
+	}
+
+	success := &pb.StorageNodeMessage{
+		Message: &pb.StorageNodeMessage_Succeeded{
+			Succeeded: &pb.TransferSucceeded{
+				OriginalPieceId:      transferPiece.OriginalPieceId,
+				OriginalPieceHash:    originalHash,
+				OriginalOrderLimit:   originalOrderLimit,
+				ReplacementPieceHash: pieceHash,
+			},
+		},
+	}
+	return c.Send(success)
```

Comment:
nit - does error need to be wrapped (idk)

Reply:
I think we wrap it at the higher level so it's fine